backend/lib/services/terminals/terminalBootstrap.js

```diff
@@ -455,44 +530,79 @@ class Bootstrapper extends Queue {
       if (this.bootstrapPending.containsResource(resource)) {
         return
       }
-      const key = this.bootstrapPending.addResource(resource)
-      logger.debug(`bootstrapping of ${key} postponed`)
+      this.bootstrapPending.addResource(resource)
+      logger.debug(`bootstrapping of ${description} postponed`)
       return
     }
 
     if (this.bootstrapPending.containsResource(resource)) {
       this.bootstrapPending.removeResource(resource)
     }
 
-    const description = `${kind} - ${resource.metadata.name}`
-    const handler = new Handler(() => {
-      switch (kind) {
-        case 'Seed':
-          return handleSeed(resource)
-        case 'Shoot':
-          return handleShoot(resource)
-        default:
-          logger.error(`can't bootstrap unsupported kind ${kind}`)
-      }
-    }, description)
+    const key = BootstrappedSet.keyForResource(resource)
+    const taskId = taskIdForResource(resource)
+    const handler = new Handler({
+      id: taskId, // with the id we make sure that the task for one shoot is not added multiple times (e.g. on another ADDED event when the shoot watch is re-established)
+      fn: async session => {
+        if (session.canceled) {
+          logger.debug(`Canceling handler of ${description} as requested`)
+          return
+        }
+
+        switch (kind) {
+          case 'Seed': {
+            await handleSeed({ name })
+            break
+          }
+          case 'Shoot': {
+            await handleShoot({ name, namespace })
+            break
+          }
+          default: {
+            logger.error(`can't bootstrap unsupported kind ${kind}`)
+            return
+          }
+        }
+
+        if (session.canceled) { // do not add key to the bootstrapped set when the session is canceled (due to shoot deletion) to prevent leaking memory
+          logger.debug(`Canceling handler of ${description} as requested after handling resource`)
+          return
+        }
+        logger.debug(`Successfully bootstrapped ${description}`)
+        this.bootstrapped.addResourceWithKey(key)
+      },
+      description
+    })
```

Comment:
```suggestion
    const fn = async session => {
        if (session.canceled) {
          logger.debug(`Canceling handler of ${description} as requested`)
          return
        }

        switch (kind) {
          case 'Seed': {
            await handleSeed({ name })
            break
          }
          case 'Shoot': {
            await handleShoot({ name, namespace })
            break
          }
          default: {
            logger.error(`can't bootstrap unsupported kind ${kind}`)
            return
          }
        }

        if (session.canceled) { // do not add key to the bootstrapped set when the session is canceled (due to shoot deletion) to prevent leaking memory
          logger.debug(`Canceling handler of ${description} as requested after handling resource`)
          return
        }
        logger.debug(`Successfully bootstrapped ${description}`)
        this.bootstrapped.addResourceWithKey(key)
      }
    const handler = new Handler(fn, {
      id: taskId, // with the id we make sure that the task for one shoot is not added multiple times (e.g. on another ADDED event when the shoot watch is re-established)
      description
    })
```